[buteo-sync-plugins-social] Commit changes after all remote communication succeeds. Contributes to MER#1248

chriadam opened the pull request:
This commit delays any mkcal database activity until after all
communication with the Google Calendar has completed successfully.
This ensures that the database remains in a usable and recoverable
state, and minimises the chance that we need to perform a clean sync
cycle (which can result in lost-local-modifications).
It includes partial-upsync detection via a private extended property.

Contributes to MER#1248

chriadam commented:
@antseppa @pvuorela @blammit this one is ready for review now.  It basically consists of two parts:

1) perform all network communication, and cache the objects/results in memory.  When everything is complete, then write to database.

2) perform partial-upsync detection.  Because we can upsync some change (eg, a local addition) and then fail (due to later network requests failing) we can get into a state where the remote reports an addition which actually began life as a local addition.  So, I now handle that via a custom extension property which is upsynced to Google.

Long and short of it is: we should not need to perform a cleansync nearly as often as in the past, so we should suffer less lost-update problems.

blammit commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -870,9 +934,12 @@ void GoogleCalendarSyncAdaptor::purgeDataForOldAccount(int oldId, SocialNetworkS
-        SOCIALD_LOG_INFO("last sync was not successful; performing clean sync");
+        SOCIALD_LOG_INFO("performing clean sync");
+    } else if (!lastSyncSuccessful) {
+        SOCIALD_LOG_INFO("last sync was not successful, attempting to recover without clean sync");
```
It looks like the bool returned by wasLastSyncSuccessful() is not used anymore, except for debugging. Is it still useful to store this value? Should the code document what are recoverable vs unrecoverable circumstances?

chriadam replied:
I wanted to have the possibility to set this value explicitly, so that we can set it in the future for:

a) error handling in some case which we currently have not considered

b) some UI setting to allow the user to perform a clean sync just in case things go pear shaped

Currently it's not used just because I can't see a case which would be completely unrecoverable using the current code, but that doesn't mean that such an unrecoverable situation does not exist.  And I'd prefer not to rewrite the .ini handling code again, if/when we hit that situation in the future.

blammit commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -1241,327 +1305,351 @@ void GoogleCalendarSyncAdaptor::finishedRequestingRemoteEvents(int accountId, co
         return; // still waiting for more requests to finish.
```
eventsFinishedHandler() may call this will updateTimestampStr=QString() in the error case and the empty string is still added to m_calendarsFinishedRequested; is that okay?

chriadam replied:
good catch... I will need to investigate this.

chriadam replied:
At line 823 of the new PR, we check to see if the m_syncSucceeded[accountId] value is true prior to attempting to use that value - so, yes, I think it's ok.

chriadam replied:
But I will add a comment here to make that clear.

blammit commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -1241,327 +1305,351 @@ void GoogleCalendarSyncAdaptor::finishedRequestingRemoteEvents(int accountId, co
+                // however the alternative is to lose remote changes made since then...
+                // So we stick with our "prefer-remote" conflict resolution strategy here.
+                SOCIALD_LOG_DEBUG("Reloading partial upsync artifact:" << eventId << "from server as a modification");
+                changed = true; // actually a no-op, the etag/gcalId comment field change will ensure this is set above.
```
The bool is not used anywhere?

blammit replied:
Why is it set if it's a no-op?

chriadam replied:
I will comment out the line but leave it in.  I prefer the code to explicitly document these sorts of decisions just because it's complicated.

blammit replied:
If it must be true in this case and the debug is required, how about 

 if (partialUpsyncArtifactsNeedingUpdate.contains(eventId)) {
    // debug
     changed = true;
} else {
     changed = remoteModificationIsReal(eventData, event);
}

If the debug is not required then
bool changed = partialUpsyncArtifactsNeedingUpdate.contains(eventId) || remoteModificationIsReal(eventData, event)

chriadam replied:
done, thanks

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -1737,43 +1821,281 @@ void GoogleCalendarSyncAdaptor::upsyncFinishedHandler()
                 m_syncSucceeded[accountId] = false;
             } else {
                 // update this event in the local calendar
+                googleNotebook->setIsReadOnly(false);
```
Why is this needed here if local changes are not done yet?

chriadam replied:
Good catch, I think it isn't.

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -1737,43 +1821,281 @@ void GoogleCalendarSyncAdaptor::upsyncFinishedHandler()
+                    } else {
+                        // this is a new event in its own right.
+                        SOCIALD_LOG_DEBUG("Event added remotely:" << eventId);
+                        event = KCalCore::Event::Ptr(new KCalCore::Event);
```
Suppose there are two or more devices running this code, event is created and upsynced from one, and then gets downsynced to rest. It would be then new event on those other devices and as mkcal uid is stored on google side, it would need to be used here as well?

chriadam replied:
Urgh. You're right, I need to consider the multiple-device case more carefully.  Perhaps I can simply set the uid on downsync.  I will try it.

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -1241,327 +1307,353 @@ void GoogleCalendarSyncAdaptor::finishedRequestingRemoteEvents(int accountId, co
 }
 
-void GoogleCalendarSyncAdaptor::updateLocalCalendarNotebookEvents(int accountId, const QString &accessToken,
-                                                                  const QString &calendarId, const QDateTime &since)
```
Declaration seems to still exist in header.

chriadam replied:
Thanks, also fixed

pvuorela commented:
Complexity for these modification lists still scares me. Out of curiosity, would there problems using event revision numbers, or last modified date, for deciding what needs to be changed or upsynced?

chriadam commented:
It should make no difference whether we use an event revision number versus an etag, right?  If the etag has changed, then the event has changed (and should result in a revision number bump).

I don't think we can use the last modified date, as the user can change the system date time and so forth, which would result in problems.

pvuorela commented:
Sure, etag should be fine. I guess I'm mostly wondering what can go wrong if changes are not first gathered and then committed, but rather per event deciding if something has changed.

chriadam commented:
I fixed the couple of review comments, and also added the "set local event uid to x-jolla-sociald-uid value if remote-added" thing -- but I haven't tested at all.  So, consider this one currently WIP.

chriadam commented:
Rebased.

Testing shows a couple of issues:

1) I currently set the x-uid field on every upsync rather than only for the first (local addition) upsync.  This results in spurious changes being detected in multi-device case.

2) the "re-use same uid" code I wrote isn't being triggered, as far as I can tell from the logs.  I need to find out why, and fix that one too.

Aside from that, it seems to work sanely - things are committed when required and the database seems to be left in sane states when connectivity drops halfway through, etc.  Multi-device cases seem to work well aside from the partial upsync problem caused by (2), with deletions and modifications being synced correctly (but of course (1) means that sometimes the remote has "changed" when it really hasn't, causing some local changes to be discarded if you sync immediately - but that will be fixed when (1) is fixed).

So, I will fix these two issues and then hopefully we'll be good to go.

chriadam commented:
I fixed (1) and it turns out that (2) works also, I was misreading it last time.  I also fixed a crash bug which could be hit in clean-sync case.

However, we still have spurious local-changes being upsynced.  The flow is this:

a) Sync, get remote event, and create it locally.
b) Next sync, we see the same event on remote, and check if the modification is real.  It's not, so we discard the "change" as spurious.
c) Then we see that event as having been locally added (due to (a)) and we currently do not detect that change as being spurious (we only detect whether or not it was previously downsynced, and in that case, we treat it as a modification rather than an addition).

So in the log, we see lines like:

Detected remote modification: "remoteEventId" in "calendarName" is real? false

and then a bit later:

Converting local addition to modification due to it being a previously downsynced event

or (in the modify case, instead of insertion):

queueing upsync modification for gcal id: "remoteEventId"

So, what I need to do is detect whether the "discarded remote non-change" is equivalent to the "current (supposedly changed) local event", and if so, discard the local change as spurious.

chriadam commented:
Ok, we now detect that case, and avoid it.

However, we STILL have spurious upsyncs occurring in the case where a previous downsync is reported on the next cycle as a local modification.  In that case, there is no remote modification reported during the second sync cycle, and so we have nothing to compare the local event to, to determine whether the change is spurious or real...

We can resolve this by simply requesting events for a longer time period, instead of using the sync timestamp as the "updated-min" field.  But that means increasing the amount of data used per sync... That's probably ok, I guess...  I'm just not sure if it's worth it, as the current code shouldn't cause data loss in any case.

chriadam commented:
Ok, this one now extends the updatedMin period back a week, to ensure that we get any modifications which occurred during the previous sync cycle.  I've also updated the delta calculation to be more precise about why things are discarded.

This is now ready for review, once more.

pvuorela commented:
I've been trying to review changes and code as a whole, and my brain starts to hurt if I try to follow everything that's happening. But since I cannot point anything obviously wrong and a few people have already been involved, I think it should be good to go. 

LGTM.

chriadam commented:
Rebased and merging.  Thanks!
